File: packages/shared/src/shared/shell_protocol.py
"""Framed wire protocol for interactive container shells.

The gateway shell tunnel (``/api/v1/shells/id/{stub_id}/{container_id}``) is a
raw bidirectional byte pipe between a shell client and the worker-owned static
container helper mounted into every OCI container. Both ends speak this
framing so control messages (auth, resize, exit) can share the pipe with PTY
bytes without requiring Python or SDK packages in the user image.

Frame layout: 1-byte frame type + 4-byte big-endian payload length + payload.

Client -> server frames: ``Auth`` (JSON ``ShellAuthRequest``, must be first),
``Data`` (stdin bytes), ``Resize`` (JSON ``ShellResizeRequest``).
Server -> client frames: ``Ready`` (auth accepted), ``Error`` (UTF-8 message),
``Data`` (PTY output bytes), ``Exit`` (UTF-8 integer exit code).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from pydantic import BaseModel, ConfigDict

SHELL_FRAME_HEADER_SIZE = 5
SHELL_FRAME_MAX_PAYLOAD_BYTES = 1024 * 1024
SHELL_DEFAULT_TERM = "xterm-256color"


class ShellFrameType(bytes, Enum):
    """Frame type bytes. ``Data`` carries stdin (client->server) or PTY
    output (server->client); the remaining types are direction-specific."""

    Auth = b"A"
    Data = b"D"
    Resize = b"R"
    Ready = b"O"
    Error = b"E"
    Exit = b"X"


class ShellAuthRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    username: str
    password: str
    term: str = SHELL_DEFAULT_TERM
    cols: int = 80
    rows: int = 24
    probe: bool = False


class ShellResizeRequest(BaseModel):
    model_config = ConfigDict(extra="ignore")

    cols: int
    rows: int


@dataclass(frozen=True, slots=True)
class ShellFrame:
    frame_type: bytes
    payload: bytes


class ShellFrameError(ValueError):
    pass


def encode_shell_frame(frame_type: bytes, payload: bytes = b"") -> bytes:
    if len(frame_type) != 1:
        msg = "shell frame type must be a single byte"
        raise ShellFrameError(msg)
    if len(payload) > SHELL_FRAME_MAX_PAYLOAD_BYTES:
        msg = "shell frame payload exceeds maximum size"
        raise ShellFrameError(msg)
    return frame_type + len(payload).to_bytes(4, "big") + payload


class ShellFrameDecoder:
    """Incremental decoder for the framed shell byte stream."""

    def __init__(self) -> None:
        self._buffer = bytearray()

    def feed(self, data: bytes) -> list[ShellFrame]:
        self._buffer.extend(data)
        frames: list[ShellFrame] = []
        while True:
            frame = self._next_frame()
            if frame is None:
                return frames
            frames.append(frame)

    def _next_frame(self) -> ShellFrame | None:
        if len(self._buffer) < SHELL_FRAME_HEADER_SIZE:
            return None
        length = int.from_bytes(self._buffer[1:SHELL_FRAME_HEADER_SIZE], "big")
        if length > SHELL_FRAME_MAX_PAYLOAD_BYTES:
            msg = "shell frame payload exceeds maximum size"
            raise ShellFrameError(msg)
        end = SHELL_FRAME_HEADER_SIZE + length
        if len(self._buffer) < end:
            return None
        frame_type = bytes(self._buffer[0:1])
        payload = bytes(self._buffer[SHELL_FRAME_HEADER_SIZE:end])
        del self._buffer[:end]
        return ShellFrame(frame_type=frame_type, payload=payload)


__all__ = [
    "SHELL_DEFAULT_TERM",
    "SHELL_FRAME_HEADER_SIZE",
    "SHELL_FRAME_MAX_PAYLOAD_BYTES",
    "ShellAuthRequest",
    "ShellFrame",
    "ShellFrameDecoder",
    "ShellFrameError",
    "ShellFrameType",
    "ShellResizeRequest",
    "encode_shell_frame",
]

File: apps/web/src/lib/shell-protocol.ts
/**
 * Browser-side codec for the framed interactive shell protocol.
 *
 * Mirrors `shared.shell_protocol` (Python). Every frame is a 1-byte type, a
 * 4-byte big-endian payload length, then the payload. The gateway WebSocket
 * tunnel forwards these bytes verbatim to the in-container shell server, so the
 * browser must authenticate before sending stdin.
 */

const HEADER_SIZE = 5;
const MAX_PAYLOAD_BYTES = 1024 * 1024;

export const ShellFrameType = {
  Auth: 0x41, // "A"
  Data: 0x44, // "D" — stdin (client->server) or PTY output (server->client)
  Resize: 0x52, // "R"
  Ready: 0x4f, // "O"
  Error: 0x45, // "E"
  Exit: 0x58, // "X"
} as const;

export type ShellFrame = { type: number; payload: Uint8Array };

const encoder = new TextEncoder();

export function encodeShellFrame(
  type: number,
  payload: Uint8Array = new Uint8Array(),
): Uint8Array<ArrayBuffer> {
  const buffer = new ArrayBuffer(HEADER_SIZE + payload.length);
  const frame = new Uint8Array(buffer);
  frame[0] = type;
  new DataView(buffer).setUint32(1, payload.length, false);
  frame.set(payload, HEADER_SIZE);
  return frame;
}

export function encodeJsonFrame(type: number, value: unknown): Uint8Array<ArrayBuffer> {
  return encodeShellFrame(type, encoder.encode(JSON.stringify(value)));
}

/** Incremental decoder that reassembles frames across WebSocket message boundaries. */
export class ShellFrameDecoder {
  private buffer: Uint8Array = new Uint8Array(0);

  feed(data: Uint8Array): ShellFrame[] {
    this.buffer = concat(this.buffer, data);
    const frames: ShellFrame[] = [];
    for (;;) {
      if (this.buffer.length < HEADER_SIZE) break;
      const length = new DataView(this.buffer.buffer, this.buffer.byteOffset + 1, 4).getUint32(
        0,
        false,
      );
      if (length > MAX_PAYLOAD_BYTES) {
        throw new Error("shell frame payload exceeds maximum size");
      }
      const end = HEADER_SIZE + length;
      if (this.buffer.length < end) break;
      frames.push({
        type: this.buffer[0]!,
        payload: this.buffer.slice(HEADER_SIZE, end),
      });
      this.buffer = this.buffer.slice(end);
    }
    return frames;
  }
}

function concat(left: Uint8Array, right: Uint8Array): Uint8Array {
  if (left.length === 0) return right;
  const out = new Uint8Array(left.length + right.length);
  out.set(left, 0);
  out.set(right, left.length);
  return out;
}
